modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientResource.java

```diff
@@ -17,20 +17,45 @@
 
 package org.apache.ignite.client.handler;
 
+import org.apache.ignite.lang.IgniteException;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
 /**
  * Client resource holder.
  */
-public interface ClientResource {
+public class ClientResource {
+    /** Object. */
+    private final Object obj;
+
+    /** Release delegate. */
+    private final @Nullable Runnable releaseRunnable;
+
+    public ClientResource(@NotNull Object obj, @Nullable Runnable releaseRunnable) {
+        this.obj = obj;
+        this.releaseRunnable = releaseRunnable;
+    }
+
     /**
      * Gets the underlying object.
      *
      * @param <T> Object type.
      * @return Object.
      */
-    <T> T get();
+    public <T> T get(Class<T> type) {
+        if (!type.isInstance(obj)) {
+            throw new IgniteException("Incorrect resource type. Expected " + type.getName() + ", but got " + obj.getClass().getName());
```

Comment:
Why can not we get type information from `T`?

Reply:
Due to Java type erasure, `T` is replaced with `Object` at compile time, so there is no type information at run time.